survey_auto_code/survey_auto_code.py

```diff
@@ -0,0 +1,127 @@
+    }
+
+    # Load data from JSON file
+    with open(demog_1_input_path, "r") as f:
+        demog_1_data = TracedDataJsonIO.import_json_to_traced_data_iterable(f)
+    with open(demog_2_input_path, "r") as f:
+        demog_2_data = TracedDataJsonIO.import_json_to_traced_data_iterable(f)
+    with open(practice_input_path, "r") as f:
+        practice_data = TracedDataJsonIO.import_json_to_traced_data_iterable(f)
+
+    def join_traced_data_iterables(join_on_key, data1, data2):
+        # TODO: Move to CoreDataModules
+        """
+        Outer-joins two iterables of TracedData on the given key.
+
+        :param join_on_key: Key to join data on
+        :type join_on_key: str
+        :param data1: TracedData items to join with data2
+        :type data1: iterable of TracedData
+        :param data2: TracedData items to join with data1
+        :type data2: iterable of TracedData
+        :return: data1 outer-joined with data2 on join_on_key
+        :rtype: list of TracedData
+        """
+        data_1_lut = {td[join_on_key]: td for td in data1}
+        data_2_lut = {td[join_on_key]: td for td in data2}
+
+        ids_in_data1_only = data_1_lut.keys() - data_2_lut.keys()
+        ids_in_data2_only = data_2_lut.keys() - data_1_lut.keys()
+        ids_in_both = set(data_1_lut.keys()).intersection(set(data_2_lut.keys()))
+
+        merged_data = []
+        for id in ids_in_data1_only:
+            merged_data.append(data_1_lut[id])
+        for id in ids_in_data2_only:
+            merged_data.append(data_2_lut[id])
+        for id in ids_in_both:
+            # TODO: Preserve history from both trees.
+            # TODO: Assert all keys in common between the two TracedData items here have identical values.
+            data_1_lut[id].append_data(
+                dict(data_2_lut[id].items()),
+                Metadata(user, Metadata.get_call_location(), time.time())
+            )
+            merged_data.append(data_1_lut[id])
+
+        return merged_data
+
+    demog_data = join_traced_data_iterables("avf_phone_id", demog_1_data, demog_2_data)
+    all_survey_data = join_traced_data_iterables("avf_phone_id", demog_data, practice_data)
+
+    # Clean the survey responses
+    for td in all_survey_data:
+        for key, cleaner in cleaning_plan.items():
+            if cleaner is not None and key in td:
+                td.append_data(
+                    {"{}_clean".format(key): cleaner(td[key])},
+                    Metadata(user, Metadata.get_call_location(), time.time())
+                )
+
+    # Set missing entries in the raw data to 'NA'
+    for td in all_survey_data:
+        for key in cleaning_plan:
+            if key not in td:
+                td.append_data({key: "NA"}, Metadata(user, Metadata.get_call_location(), time.time()))
+
+    # Write json output
+    if os.path.dirname(json_output_path) is not "" and not os.path.exists(os.path.dirname(json_output_path)):
```

Comment:
I thought we already had a helper for this?

Reply:
We will once #12 is merged